forecasttools/idata_w_dates_to_df.py

```diff
@@ -0,0 +1,127 @@
+"""
+Contains functions for converting
+Arviz InferenceData objects to Polars
+dataframes with dates and draws and
+for working with conversions of
+Polars dataframes to Tidy dataframes.
+"""
+
+from datetime import datetime
+
+import arviz as az
+import numpy as np
+import polars as pl
+
+
+def add_dates_as_coords_to_idata(
+    idata_wo_dates: az.InferenceData,
+    group_dim_dict: dict[str, str],
+    start_date_iso: str,
+) -> az.InferenceData:
+    """
+    Modifies the provided idata object by assigning
+    date arrays to each group specified in group_dim_dict.
+
+    Parameters
+    ----------
+    idata
+        The InferenceData object that contains
+        multiple groups (e.g., observed_data,
+        posterior_predictive).
+    group_dim_dict
+        A dictionary that maps InferenceData group
+        names (e.g., "observed_data", "posterior_predictive"
+        ) to dimension names (e.g., "obs_dim_0").
+    start_date_iso
+        The start date in ISO format (YYYY-MM-DD) from
+        which to begin the date range for the dimension.
+
+    Returns
+    -------
+    idata
+        The modified InferenceData object with
+        date coordinates assigned to each group.
+    """
+
+    # NOTE: failure mode is if groups don't have the same
+    # start date
+
+    # convert received start date string to datetime object
+    start_date_as_dt = datetime.strptime(start_date_iso, "%Y-%m-%d")
+    # copy idata object to avoid modifying the original
+    idata_w_dates = idata_wo_dates.copy()
+    # modify indices of each selected group to dates
+    for (
+        group_name,
+        dim_name,
+    ) in group_dim_dict.items():
+        idata_group = getattr(idata_w_dates, group_name, None)
+        if idata_group is not None:
+            interval_size = idata_group.sizes[dim_name]
+            interval_dates = (
+                pl.DataFrame()
+                .select(
+                    pl.date_range(
+                        start=start_date_as_dt,
+                        end=start_date_as_dt
+                        + pl.duration(days=interval_size - 1),
+                        interval="1d",
+                        closed="both",
+                    )
+                )
+                .to_series()
```

Comment:
Seems like there should be a much more concise way to do this than creating and empty data frame, converting it to a series, and converting that to a numpy array.

Reply:
This is what I originally had and thought would work:

```python
interval_size = idata_group.sizes[dim_name]
interval_dates = pl.date_range(
    start=start_date_as_dt,
    end=start_date_as_dt + pl.duration(days=interval_size - 1),
    interval="1d",
    closed="both"
).to_numpy().astype("datetime64[ns]")
```

and this is the error I got

```
AttributeError: 'Expr' object has no attribute 'to_numpy'
```

Perhaps I am using [date_range](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.date_range.html) incorrectly.

Reply:
Silly mistake on my end. Just re-read the documentation. `eager=True` gets me what I want.

```
eager: Evaluate immediately and return a Series. If set to False (default), return an expression instead.
```